File: firmware/tools/esp32_terminal.py
#!/usr/bin/env python3
"""
ESP32 Bluetooth & Serial Terminal
Interactive terminal for ESP32-SD-WiFi device with dual monitoring

Requirements:
    pip install bleak textual
    platformio (for serial monitoring)

Usage:
    python esp32_terminal.py [device_name]
    
    device_name: Optional Bluetooth device name (default: digicam-001)
    
Examples:
    python esp32_terminal.py                    # Connect to digicam-001
    python esp32_terminal.py ESP32-SD-WiFi      # Connect to ESP32-SD-WiFi
    python esp32_terminal.py my-device          # Connect to my-device

Commands:
    HELP                              - Show available commands
    STATUS                            - Show device status
    WIFI ON/OFF                       - Enable/disable WiFi
    WIFI SCAN                         - Scan for networks
    WIFI CONNECT <ssid> <password>    - Connect to WiFi
    WIFI AP                           - Start Access Point mode
    SLEEP                             - Enter sleep mode
    WAKE                              - Wake from sleep mode
    RESTART                           - Restart device
"""

import asyncio
import sys
import subprocess
import threading
from datetime import datetime
from collections import deque
from bleak import BleakClient, BleakScanner
from textual.app import App, ComposeResult
from textual.containers import Container, Horizontal, Vertical
from textual.widgets import Header, Footer, Static, Input, RichLog
from textual.binding import Binding

# BLE UART Service UUIDs
SERVICE_UUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
RX_UUID = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"  # Write to ESP32
TX_UUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"  # Receive from ESP32

# Defaults
DEFAULT_DEVICE_NAME = "digicam-001"
DEFAULT_BAUD_RATE = 115200
MAX_LOG_LINES = 500

class ESP32Terminal(App):
    """Textual app for ESP32 terminal"""
    
    CSS = """
    #bt_log {
        border: solid cyan;
        height: 100%;
    }
    
    #serial_log {
        border: solid magenta;
        height: 100%;
    }
    
    #logs_container {
        height: 1fr;
    }
    
    #input_container {
        height: auto;
        padding: 1;
    }
    """
    
    BINDINGS = [
        Binding("ctrl+c", "quit", "Quit"),
    ]
    
    def __init__(self, device_name=DEFAULT_DEVICE_NAME):
        super().__init__()
        # Bluetooth
        self.client = None
        self.bt_connected = False
        self.device_name = device_name
        
        # Serial (via platformio)
        self.serial_process = None
        self.serial_connected = False
        
        # UI state
        self.running = False
        self.bt_log_widget = None
        self.serial_log_widget = None
        self.input_widget = None
        
        # Message queues for thread-safe logging
        self.bt_message_queue = asyncio.Queue()
        self.serial_message_queue = asyncio.Queue()
        
        # Buffers for periodic updates (reduces render frequency)
        self.bt_buffer = []
        self.serial_buffer = []
        
    def compose(self) -> ComposeResult:
        """Create child widgets"""
        yield Header()
        with Vertical(id="logs_container"):
            with Horizontal():
                yield RichLog(id="bt_log", highlight=True, markup=True, max_lines=1000)
                yield RichLog(id="serial_log", highlight=True, markup=True, max_lines=1000)
        with Container(id="input_container"):
            yield Input(placeholder="Type command and press Enter (or 'quit' to exit)")
        yield Footer()
    
    def on_mount(self) -> None:
        """Called when app starts"""
        self.bt_log_widget = self.query_one("#bt_log", RichLog)
        self.serial_log_widget = self.query_one("#serial_log", RichLog)
        self.input_widget = self.query_one(Input)
        
        self.bt_log_widget.border_title = "Bluetooth"
        self.serial_log_widget.border_title = "Serial Monitor"
        
        self.input_widget.focus()
        
        # Get event loop for thread-safe operations
        self.loop = asyncio.get_event_loop()
        
        # Start connections
        self.running = True
        asyncio.create_task(self.connect_devices())
        
        # Start message collectors (collect messages into buffers)
        asyncio.create_task(self.collect_bt_messages())
        asyncio.create_task(self.collect_serial_messages())
        
        # Start periodic renderers (update UI on timer, not on every message)
        self.set_interval(0.1, self.render_bt_buffer)  # Update BT every 100ms
        self.set_interval(0.2, self.render_serial_buffer)  # Update Serial every 200ms
    
    async def connect_devices(self):
        """Connect to Bluetooth and start serial monitor"""
        await self.connect_bluetooth()
        self.start_serial_monitor()
        
        if self.bt_connected:
            await self.send_command("HELP")
    
    def notification_handler(self, sender, data):
        """Handle incoming data from ESP32 via Bluetooth"""
        try:
            message = data.decode('utf-8').rstrip('\n\r')
            if message:
                timestamp = datetime.now().strftime("%H:%M:%S")
                # Queue message for processing in main loop
                asyncio.run_coroutine_threadsafe(
                    self.bt_message_queue.put(f"[dim]{timestamp}[/] {message}"),
                    self.loop
                )
        except UnicodeDecodeError:
            timestamp = datetime.now().strftime("%H:%M:%S")
            asyncio.run_coroutine_threadsafe(
                self.bt_message_queue.put(f"[dim]{timestamp}[/] [yellow][Binary: {data.hex()}][/]"),
                self.loop
            )
    
    async def collect_bt_messages(self):
        """Collect Bluetooth messages into buffer without rendering"""
        while self.running:
            try:
                message = await self.bt_message_queue.get()
                self.bt_buffer.append(message)
                # Limit buffer size
                if len(self.bt_buffer) > 100:
                    self.bt_buffer = self.bt_buffer[-100:]
            except Exception:
                pass
    
    async def collect_serial_messages(self):
        """Collect Serial messages into buffer without rendering"""
        while self.running:
            try:
                message = await self.serial_message_queue.get()
                self.serial_buffer.append(message)
                # Limit buffer size
                if len(self.serial_buffer) > 100:
                    self.serial_buffer = self.serial_buffer[-100:]
            except Exception:
                pass
    
    def render_bt_buffer(self):
        """Render buffered BT messages (called on timer)"""
        if self.bt_buffer:
            with self.batch_update():
                for msg in self.bt_buffer:
                    self.bt_log_widget.write(msg)
            self.bt_buffer = []
    
    def render_serial_buffer(self):
        """Render buffered Serial messages (called on timer)"""
        if self.serial_buffer:
            with self.batch_update():
                for msg in self.serial_buffer:
                    self.serial_log_widget.write(msg)
            self.serial_buffer = []
    
    async def connect_bluetooth(self):
        """Scan for and connect to ESP32 device via Bluetooth"""
        self.bt_log_widget.write(f"[yellow]Scanning for '{self.device_name}'...[/]")
        device = await BleakScanner.find_device_by_name(self.device_name, timeout=10.0)
        
        if not device:
            self.bt_log_widget.write(f"[red]❌ Device '{self.device_name}' not found![/]")
            return False
        
        self.bt_log_widget.write(f"[green]✓ Found: {device.name} ({device.address})[/]")
        
        try:
            self.client = BleakClient(device)
            await self.client.connect()
            await self.client.start_notify(TX_UUID, self.notification_handler)
            
            self.bt_connected = True
            self.bt_log_widget.write("[green bold]✓ Bluetooth connected![/]")
            self.sub_title = "BT: Connected"
            return True
            
        except Exception as e:
            self.bt_log_widget.write(f"[red]❌ Connection failed: {e}[/]")
            return False
    
    def start_serial_monitor(self):
        """Start platformio serial monitor"""
        try:
            self.serial_log_widget.write("[yellow]Starting platformio serial monitor...[/]")
            self.serial_process = subprocess.Popen(
                ['platformio', 'device', 'monitor', '--baud', '115200'],
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                bufsize=1,
                universal_newlines=True
            )
            self.serial_connected = True
            self.serial_log_widget.write("[green bold]✓ Serial monitor started![/]")
            
            # Start reading in thread
            serial_thread = threading.Thread(target=self.read_serial_thread, daemon=True)
            serial_thread.start()
            return True
        except FileNotFoundError:
            self.serial_log_widget.write("[red]❌ platformio not found. Install it first.[/]")
            return False
        except Exception as e:
            self.serial_log_widget.write(f"[red]❌ Failed to start serial monitor: {e}[/]")
            return False
    
    def read_serial_thread(self):
        """Read from platformio serial monitor in thread with throttling"""
        import time
        last_update = time.time()
        buffer = []
        
        while self.running and self.serial_connected and self.serial_process:
            try:
                line = self.serial_process.stdout.readline()
                if line:
                    message = line.rstrip('\n\r')
                    if message and not message.startswith('---'):  # Skip platformio headers
                        buffer.append(message)
                        
                        # Throttle: only send to queue every 100ms or when buffer has 20 lines
                        current_time = time.time()
                        if (current_time - last_update) >= 0.1 or len(buffer) >= 20:
                            timestamp = datetime.now().strftime("%H:%M:%S")
                            # Send all buffered lines as one message
                            combined = "\n".join(buffer)
                            asyncio.run_coroutine_threadsafe(
                                self.serial_message_queue.put(f"[dim]{timestamp}[/] {combined}"),
                                self.loop
                            )
                            buffer = []
                            last_update = current_time
                            
                elif self.serial_process.poll() is not None:
                    # Flush remaining buffer
                    if buffer:
                        timestamp = datetime.now().strftime("%H:%M:%S")
                        combined = "\n".join(buffer)
                        asyncio.run_coroutine_threadsafe(
                            self.serial_message_queue.put(f"[dim]{timestamp}[/] {combined}"),
                            self.loop
                        )
                    
                    # Process ended
                    self.serial_connected = False
                    timestamp = datetime.now().strftime("%H:%M:%S")
                    asyncio.run_coroutine_threadsafe(
                        self.serial_message_queue.put(f"[dim]{timestamp}[/] [yellow]Serial monitor stopped[/]"),
                        self.loop
                    )
                    break
            except Exception as e:
                timestamp = datetime.now().strftime("%H:%M:%S")
                asyncio.run_coroutine_threadsafe(
                    self.serial_message_queue.put(f"[dim]{timestamp}[/] [red]Error: {e}[/]"),
                    self.loop
                )
    
    async def send_command(self, command):
        """Send command to ESP32 via Bluetooth"""
        if not self.bt_connected or not self.client:
            timestamp = datetime.now().strftime("%H:%M:%S")
            self.bt_log_widget.write(f"[dim]{timestamp}[/] [red]Not connected![/]")
            return False
        
        try:
            timestamp = datetime.now().strftime("%H:%M:%S")
            self.bt_log_widget.write(f"[dim]{timestamp}[/] [green]> {command}[/]")
            await self.client.write_gatt_char(RX_UUID, command.encode('utf-8'))
            
            # Don't wait for response - let it come asynchronously
            # The notification handler will display the response
            
            return True
        except Exception as e:
            timestamp = datetime.now().strftime("%H:%M:%S")
            self.bt_log_widget.write(f"[dim]{timestamp}[/] [red]Error: {e}[/]")
            return False
    
    def on_input_submitted(self, event: Input.Submitted) -> None:
        """Handle command submission"""
        command = event.value.strip()
        self.input_widget.value = ""
        
        if command.lower() in ['quit', 'exit', 'q']:
            self.exit()
            return
        
        if command and self.bt_connected:
            # Run in background to not block UI
            asyncio.create_task(self.send_command(command))
    
    async def on_unmount(self) -> None:
        """Called when app is closing"""
        self.running = False
        
        if self.client and self.bt_connected:
            await self.client.disconnect()
            self.bt_connected = False
        
        if self.serial_process and self.serial_connected:
            self.serial_process.terminate()
            try:
                self.serial_process.wait(timeout=2)
            except subprocess.TimeoutExpired:
                self.serial_process.kill()
            self.serial_connected = False
    

    


def main():
    # Get device name from command line argument or use default
    device_name = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_DEVICE_NAME
    
    # Create and run the app
    app = ESP32Terminal(device_name)
    app.title = "ESP32 Terminal"
    app.sub_title = f"Device: {device_name}"
    app.run()

if __name__ == "__main__":
    main()

File: firmware/tools/esp32_terminal.js
#!/usr/bin/env node
/**
 * ESP32 Bluetooth & Serial Terminal
 * Node.js version with blessed TUI
 */

const blessed = require('blessed');
const noble = require('@abandonware/noble');
const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');
const pty = require('node-pty');

// BLE UART Service UUIDs
const SERVICE_UUID = '6e400001b5a3f393e0a9e50e24dcca9e';
const RX_UUID = '6e400002b5a3f393e0a9e50e24dcca9e';
const TX_UUID = '6e400003b5a3f393e0a9e50e24dcca9e';

const DEFAULT_DEVICE_NAME = 'digicam-001';

class ESP32Terminal {
  constructor(deviceName = DEFAULT_DEVICE_NAME) {
    this.deviceName = deviceName;
    this.blePeripheral = null;
    this.bleCharacteristics = {};
    this.serialProcess = null;
    this.screen = null;
    this.btLog = null;
    this.serialLog = null;
    this.input = null;
    this.statusBox = null;
    
    // Serial data buffer for handling \r properly
    this.serialBuffer = '';
    this.serialFlushTimer = null;
    
    // ESP32 status data
    this.esp32Status = {
      power: 'Unknown',
      wifi: 'Unknown',
      ap: 'Unknown',
      ip: 'Unknown',
      lastUpdate: null
    };
    
    // Log file setup
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-').split('T')[0];
    const logDir = path.join(__dirname, 'logs');
    if (!fs.existsSync(logDir)) {
      fs.mkdirSync(logDir, { recursive: true });
    }
    this.btLogFile = path.join(logDir, `bt_${timestamp}.txt`);
    this.serialLogFile = path.join(logDir, `serial_${timestamp}.txt`);
    
    // Create log files
    fs.writeFileSync(this.btLogFile, `ESP32 Bluetooth Log - ${new Date().toISOString()}\n${'='.repeat(60)}\n\n`);
    fs.writeFileSync(this.serialLogFile, `ESP32 Serial Log - ${new Date().toISOString()}\n${'='.repeat(60)}\n\n`);
  }

  createUI() {
    // Create screen
    this.screen = blessed.screen({
      smartCSR: true,
      title: 'ESP32 Terminal',
      mouse: true,  // Enable mouse support
      sendFocus: true,
      warnings: true
    });
    
    // Enable raw mode to capture all keys including Ctrl-C
    this.screen.program.key('C-c', () => {
      this.exit();
    });

    // Header
    const header = blessed.box({
      top: 0,
      left: 0,
      width: '100%',
      height: 3,
      content: '{center}{bold}ESP32 Bluetooth & Serial Terminal{/bold}{/center}',
      tags: true,
      style: {
        fg: 'white',
        bg: 'blue',
        bold: true
      }
    });

    // Status panel
    this.statusBox = blessed.box({
      top: 3,
      left: 0,
      width: '100%',
      height: 5,
      border: { type: 'line' },
      label: ' ESP32 Status ',
      tags: true,
      style: {
        fg: 'white',
        border: { fg: 'yellow' }
      }
    });

    // Bluetooth log (left pane)
    this.btLog = blessed.log({
      top: 8,
      left: 0,
      width: '50%',
      height: '100%-11',
      border: { type: 'line' },
      label: ' Bluetooth ',
      tags: true,
      scrollable: true,
      alwaysScroll: true,
      mouse: true,  // Enable mouse for this widget
      keys: true,   // Enable keyboard scrolling
      vi: true,     // Enable vi-style scrolling (j/k)
      scrollbar: {
        ch: ' ',
        style: { bg: 'cyan' }
      },
      style: {
        fg: 'cyan',
        border: { fg: 'cyan' }
      }
    });

    // Serial log (right pane)
    this.serialLog = blessed.log({
      top: 8,
      left: '50%',
      width: '50%',
      height: '100%-11',
      border: { type: 'line' },
      label: ' Serial Monitor ',
      tags: true,
      scrollable: true,
      alwaysScroll: true,
      mouse: true,  // Enable mouse for this widget
      keys: true,   // Enable keyboard scrolling
      vi: true,     // Enable vi-style scrolling (j/k)
      scrollbar: {
        ch: ' ',
        style: { bg: 'magenta' }
      },
      style: {
        fg: 'white',
        border: { fg: 'magenta' }
      }
    });

    // Input box
    this.input = blessed.textbox({
      bottom: 0,
      left: 0,
      width: '100%',
      height: 3,
      border: { type: 'line' },
      label: ' Command (type and press Enter, or "q/quit/esc/ctrl+c" to exit) ',
      style: {
        fg: 'white',
        border: { fg: 'green' }
      },
      inputOnFocus: true
    });

    // Append to screen
    this.screen.append(header);
    this.screen.append(this.statusBox);
    this.screen.append(this.btLog);
    this.screen.append(this.serialLog);
    this.screen.append(this.input);
    
    // Initial status display
    this.updateStatusDisplay();

    // Focus input
    this.input.focus();

    // Handle input submission
    this.input.on('submit', (value) => {
      const command = value.trim();
      this.input.clearValue();
      
      if (command.toLowerCase() === 'quit' || command.toLowerCase() === 'exit' || command.toLowerCase() === 'q') {
        this.exit();
        return;
      }
      
      if (command) {
        this.sendBLECommand(command);
      }
      
      this.input.focus();
      this.screen.render();
    });

    // Quit on Escape, q, or Ctrl-C
    this.screen.key(['escape', 'C-c', 'q'], (ch, key) => {
      this.exit();
    });

    // Also bind to raw Ctrl-C
    this.screen.key(['C-c'], (ch, key) => {
      this.exit();
    });

    this.screen.render();
  }

  updateStatusDisplay() {
    const status = this.esp32Status;
    const lastUpdate = status.lastUpdate ? status.lastUpdate.toLocaleTimeString() : 'Never';
    const btStatus = this.blePeripheral ? '{green-fg}Connected{/green-fg}' : '{red-fg}Disconnected{/red-fg}';
    
    const content = 
      `  {yellow-fg}BLE:{/yellow-fg} ${btStatus}     ` +
      `{yellow-fg}WiFi:{/yellow-fg} ${status.wifi}     ` +
      `{yellow-fg}AP:{/yellow-fg} ${status.ap}\n` +
      `  {yellow-fg}IP:{/yellow-fg} ${status.ip}     ` +
      `{yellow-fg}Power:{/yellow-fg} ${status.power}     ` +
      `{dim}Updated: ${lastUpdate}{/dim}`;
    
    this.statusBox.setContent(content);
    this.screen.render();
  }

  parseStatusResponse(message) {
    // Parse STATUS command responses to update status display
    let updated = false;
    
    // WiFi status - look for "WiFi: ON/OFF" or "WiFi ON/OFF"
    if (/wifi[:\s]+(on|off|enabled|disabled|connected)/i.test(message)) {
      const match = message.match(/wifi[:\s]+(on|off|enabled|disabled|connected)/i);
      if (match) {
        this.esp32Status.wifi = match[1].toUpperCase();
        updated = true;
      }
    }
    
    // AP/SSID - look for "AP: name" or "SSID: name"
    if (/(?:ap|ssid)[:\s]+([^\s,\n]+)/i.test(message)) {
      const match = message.match(/(?:ap|ssid)[:\s]+([^\s,\n]+)/i);
      if (match) {
        this.esp32Status.ap = match[1];
        updated = true;
      }
    }
    
    // IP address
    if (/ip[:\s]+([\d.]+)/i.test(message)) {
      const match = message.match(/ip[:\s]+([\d.]+)/i);
      if (match) {
        this.esp32Status.ip = match[1];
        updated = true;
      }
    }
    
    // Power/Battery/Voltage
    if (/([\d.]+)\s*(v|mw|ma|mah|%)/i.test(message)) {
      const match = message.match(/([\d.]+)\s*(v|mw|ma|mah|%)/i);
      if (match) {
        this.esp32Status.power = match[0];
        updated = true;
      }
    }
    
    if (updated) {
      this.esp32Status.lastUpdate = new Date();
      this.updateStatusDisplay();
    }
  }

  exit() {
    // Destroy screen first to restore terminal
    if (this.screen) {
      this.screen.destroy();
    }
    this.cleanup();
    process.exit(0);
  }

  async connectBluetooth() {
    return new Promise((resolve, reject) => {
      const timestamp = new Date().toLocaleTimeString();
      const scanMsg = `[${timestamp}] Scanning for '${this.deviceName}'...`;
      this.btLog.log(`{yellow-fg}${scanMsg}{/yellow-fg}`);
      fs.appendFileSync(this.btLogFile, scanMsg + '\n');
      this.screen.render();

      let connected = false;
      let timeoutHandle = null;

      const onDiscover = async (peripheral) => {
        if (peripheral.advertisement.localName === this.deviceName) {
          noble.stopScanning();
          
          // Clear timeout since we found the device
          if (timeoutHandle) {
            clearTimeout(timeoutHandle);
            timeoutHandle = null;
          }
          
          const timestamp = new Date().toLocaleTimeString();
          const foundMsg = `[${timestamp}] ✓ Found: ${peripheral.advertisement.localName}`;
          this.btLog.log(`{green-fg}${foundMsg}{/green-fg}`);
          fs.appendFileSync(this.btLogFile, foundMsg + '\n');
          this.screen.render();

          try {
            await peripheral.connectAsync();
            this.blePeripheral = peripheral;
            
            const { characteristics } = await peripheral.discoverSomeServicesAndCharacteristicsAsync(
              [SERVICE_UUID],
              [RX_UUID, TX_UUID]
            );

            characteristics.forEach(char => {
              this.bleCharacteristics[char.uuid] = char;
            });

            // Subscribe to notifications
            const txChar = this.bleCharacteristics[TX_UUID];
            if (txChar) {
              await txChar.subscribeAsync();
              txChar.on('data', (data) => {
                const message = data.toString('utf-8').trim();
                if (message) {
                  const timestamp = new Date().toLocaleTimeString();
                  this.btLog.log(`{cyan-fg}[${timestamp}] ${message}{/cyan-fg}`);
                  // Save to log file
                  fs.appendFileSync(this.btLogFile, `[${timestamp}] ${message}\n`);
                  
                  // Parse for status information
                  this.parseStatusResponse(message);
                  
                  this.screen.render();
                }
              });
            }

            const timestamp = new Date().toLocaleTimeString();
            const connectedMsg = `[${timestamp}] ✓ Bluetooth connected!`;
            this.btLog.log(`{green-fg}{bold}${connectedMsg}{/bold}{/green-fg}`);
            fs.appendFileSync(this.btLogFile, connectedMsg + '\n');
            
            // Update status display to show BLE connected
            this.updateStatusDisplay();
            
            this.screen.render();
            
            connected = true;
            noble.removeListener('discover', onDiscover);
            resolve(true);
          } catch (err) {
            const timestamp = new Date().toLocaleTimeString();
            const errorMsg = `[${timestamp}] ❌ Connection failed: ${err.message}`;
            this.btLog.log(`{red-fg}${errorMsg}{/red-fg}`);
            fs.appendFileSync(this.btLogFile, errorMsg + '\n');
            this.screen.render();
            
            noble.removeListener('discover', onDiscover);
            reject(err);
          }
        }
      };

      noble.on('discover', onDiscover);
      noble.startScanning([], false);

      // Timeout after 10 seconds
      timeoutHandle = setTimeout(() => {
        if (!connected) {
          noble.stopScanning();
          noble.removeListener('discover', onDiscover);
          const timestamp = new Date().toLocaleTimeString();
          const timeoutMsg = `[${timestamp}] ❌ Device '${this.deviceName}' not found!`;
          this.btLog.log(`{red-fg}${timeoutMsg}{/red-fg}`);
          fs.appendFileSync(this.btLogFile, timeoutMsg + '\n');
          this.screen.render();
          reject(new Error('Device not found'));
        }
      }, 10000);
    });
  }

  async sendBLECommand(command) {
    if (!this.blePeripheral || !this.bleCharacteristics[RX_UUID]) {
      const timestamp = new Date().toLocaleTimeString();
      const msg = `[${timestamp}] Not connected!`;
      this.btLog.log(`{red-fg}${msg}{/red-fg}`);
      fs.appendFileSync(this.btLogFile, msg + '\n');
      this.screen.render();
      return;
    }

    try {
      const timestamp = new Date().toLocaleTimeString();
      const msg = `[${timestamp}] > ${command}`;
      this.btLog.log(`{green-fg}${msg}{/green-fg}`);
      fs.appendFileSync(this.btLogFile, msg + '\n');
      
      const rxChar = this.bleCharacteristics[RX_UUID];
      await rxChar.writeAsync(Buffer.from(command, 'utf-8'), false);
      
      this.screen.render();
    } catch (err) {
      const timestamp = new Date().toLocaleTimeString();
      const msg = `[${timestamp}] Error: ${err.message}`;
      this.btLog.log(`{red-fg}${msg}{/red-fg}`);
      fs.appendFileSync(this.btLogFile, msg + '\n');
      this.screen.render();
    }
  }

  startSerialMonitor() {
    const timestamp = new Date().toLocaleTimeString();
    const startMsg = `[${timestamp}] Starting platformio serial monitor...`;
    this.serialLog.log(`{yellow-fg}${startMsg}{/yellow-fg}`);
    fs.appendFileSync(this.serialLogFile, startMsg + '\n');
    this.screen.render();

    try {
      // Use PTY to provide a pseudo-terminal for platformio
      this.serialProcess = pty.spawn('platformio', ['device', 'monitor', '--baud', '115200'], {
        name: 'xterm-color',
        cols: 80,
        rows: 30,
        cwd: process.cwd(),
        env: process.env,
        handleFlowControl: false  // Don't let PTY handle Ctrl-C
      });

      // Handle data from PTY
      this.serialProcess.onData((data) => {
        // Remove ANSI escape codes first
        const cleaned = data.replace(/\x1b\[[0-9;]*m/g, '');
        
        // Add to buffer
        this.serialBuffer += cleaned;
        
        // Clear existing flush timer
        if (this.serialFlushTimer) {
          clearTimeout(this.serialFlushTimer);
        }
        
        // Look for complete lines (ending with \n or \r\n)
        let processedSomething = false;
        while (this.serialBuffer.includes('\n')) {
          const newlineIndex = this.serialBuffer.indexOf('\n');
          let line = this.serialBuffer.substring(0, newlineIndex);
          this.serialBuffer = this.serialBuffer.substring(newlineIndex + 1);
          
          // Remove any trailing \r
          line = line.replace(/\r$/, '').trim();
          
          if (line && !line.startsWith('---')) {
            const timestamp = new Date().toLocaleTimeString();
            this.serialLog.log(`{cyan-fg}[${timestamp}] ${line}{/cyan-fg}`);
            fs.appendFileSync(this.serialLogFile, `[${timestamp}] ${line}\n`);
            processedSomething = true;
          }
        }
        
        // Set timer to flush incomplete buffer after 200ms of no new data
        this.serialFlushTimer = setTimeout(() => {
          if (this.serialBuffer.trim()) {
            const line = this.serialBuffer.trim();
            if (line && !line.startsWith('---')) {
              const timestamp = new Date().toLocaleTimeString();
              this.serialLog.log(`{cyan-fg}[${timestamp}] ${line}{/cyan-fg}`);
              fs.appendFileSync(this.serialLogFile, `[${timestamp}] ${line}\n`);
              this.screen.render();
            }
            this.serialBuffer = '';
          }
        }, 200);
        
        if (processedSomething) {
          this.screen.render();
        }
      });

      this.serialProcess.onExit(({ exitCode, signal }) => {
        const timestamp = new Date().toLocaleTimeString();
        const msg = `[${timestamp}] Serial monitor stopped (exit code: ${exitCode}, signal: ${signal})`;
        this.serialLog.log(`{yellow-fg}${msg}{/yellow-fg}`);
        fs.appendFileSync(this.serialLogFile, msg + '\n');
        this.screen.render();
      });

      const successTimestamp = new Date().toLocaleTimeString();
      const successMsg = `[${successTimestamp}] ✓ Serial monitor started!`;
      this.serialLog.log(`{green-fg}{bold}${successMsg}{/bold}{/green-fg}`);
      fs.appendFileSync(this.serialLogFile, successMsg + '\n');
      this.screen.render();
    } catch (err) {
      const timestamp = new Date().toLocaleTimeString();
      const errorMsg = `[${timestamp}] ❌ Failed to start serial monitor: ${err.message}`;
      this.serialLog.log(`{red-fg}${errorMsg}{/red-fg}`);
      fs.appendFileSync(this.serialLogFile, errorMsg + '\n');
      this.screen.render();
    }
  }

  cleanup() {
    if (this.serialProcess) {
      try {
        this.serialProcess.kill();
      } catch (err) {
        // Ignore errors on cleanup
      }
    }
    if (this.blePeripheral) {
      try {
        this.blePeripheral.disconnect();
      } catch (err) {
        // Ignore errors on cleanup
      }
    }
    
    // Print log file locations
    console.log('\nLogs saved to:');
    console.log(`  Bluetooth: ${this.btLogFile}`);
    console.log(`  Serial:    ${this.serialLogFile}`);
  }

  async start() {
    this.createUI();
    
    try {
      await this.connectBluetooth();
    } catch (err) {
      // Continue even if BT fails
    }

    this.startSerialMonitor();
  }
}

// Main
const deviceName = process.argv[2] || DEFAULT_DEVICE_NAME;
const terminal = new ESP32Terminal(deviceName);

// Handle Ctrl-C at process level
let exiting = false;
process.on('SIGINT', () => {
  if (exiting) return;
  exiting = true;
  
  if (terminal.screen) {
    terminal.exit();
  } else {
    process.exit(0);
  }
});

process.on('SIGTERM', () => {
  if (exiting) return;
  exiting = true;
  
  if (terminal.screen) {
    terminal.exit();
  } else {
    process.exit(0);
  }
});

// Catch uncaught exceptions
process.on('uncaughtException', (err) => {
  if (terminal.screen) {
    terminal.screen.destroy();
  }
  console.error('Error:', err.message);
  process.exit(1);
});

terminal.start();
